File: complete_search/78_subsets.py
class Solution:
    def subsets(self, nums: List[int]) -> List[List[int]]:
        # using recursion
        res = []
        subset = []
        def search(n: int):
            if (n == len(nums)):
                temp = []
                for idx in subset:
                    temp.append(nums[idx])
                res.append(temp)
                return
            else:
                search(n + 1)
                subset.append(n)
                search(n + 1)
                subset.pop()
        search(0)
        return res

        # using bit masking
        # res = []
        # n = len(nums)

        # for i in range(2 ** n):
        #     subset = []
        #     for j in range(n):
        #         if (i & 1 << j != 0):
        #             subset.append(nums[j])
        #     res.append(subset)
        # return res
        

File: backtracking/78_subsets.java
class Solution {
    int n = 0;
    int k;
    int[] nums = null;
    List<List<Integer>> res = new ArrayList();

    public void backtrack(int ptr, List<Integer> curr) {
        if (curr.size() == k) {
            res.add(new ArrayList(curr));
            return;
        }
        for (int i = ptr; i < n; i++) {
            curr.add(this.nums[i]);
            backtrack(i + 1, curr);
            curr.remove(curr.size() - 1);
        }
    }
    public List<List<Integer>> subsets(int[] nums) {
        n = nums.length;
        this.nums = nums;

        for (k = 0; k < n + 1; k++) {
            backtrack(0, new ArrayList<Integer>());
        }
        return res;
    }
}